controller/controller.go

```diff
@@ -239,7 +250,10 @@ func (c *Controller) delete(obj Object) {
 }
 
 func (c *Controller) propagate(resourceEvents []ResourceEvent) {
-	topology := c.topology.Build(c.cache.List())
+	topology, err := c.topology.Build(c.cache.List())
+	if !c.topology.allowTopologyLoops && err != nil {
+		panic(err)
```

Comment:
panicking does not seem to be a good idea. The controller will die again and again until the topology changes. Maybe it is better to drop an `ERR` log line + generate an event and early return waiting for a topology change.

Reply:
Not wrong. In fact, this may not be the only place where it could panic. Something very bad can also emerge from one of the dependencies underneath, beyond our control.

Ideally, any error should the handled and still move the resources to a final state, even if that "final state" (until a new event happens) consists of simply updating the status of some top-level resource to "failed".

Reply:
the fact that other places could panic is not a good reason for doing that. 

What I am asking is to handle the error

Reply:
I propose then modifying the [`ReconcileFunc`](https://github.com/Kuadrant/policy-machinery/blob/85837a3f6a5ae4b84594de2124c5338f678a57f3/controller/controller.go#L63) so it accepts propagating an error. I.e.:

```go
func(context.Context, []ResourceEvent, *machinery.Topology, error)
```

And let the user define what to do inside the registerred reconcile func.

I think this is the best approach in the end. The reconcile func could, e.g., decide to update the status of a resource informing about the error and not proceed with the rest what would be otherwise a normal reconciliation.

Neither the controller, nor the topology builder would panic.

Reply:
I am not in agreement that this should not panic. If this is the right location to panic, maybe not. But waiting of the topology to change would only make the controller flaky. The only way loops should ever be able to get into the topology is by the user defined link functions. If the user is expecting no loops they will write their controller to work with no loops, but if there is an error that they are expected to handle then their code has to be able to handle loops. Which defeats the point of have the DAG enforced in the first place.

Reply:
I understand users are in control of their linking functions and, as a general rule of thumb, should write the linking functions in ways to avoid them producing loops, if indeed they want their topologies to be a DAG. However, IMO panicking is still a harsh response. It crashes the pod, possibly before the user has the chance to inspect the broken topology.

In complex topologies, the conditions that produce a loop may be hidden deep down inside edge cases that could trigger only rarely. I can also picture these conditions emerging from transitory, self-recovering states. My point being that a system that works perfectly fine almost all the time but still occasionally crashes under very specific edge cases can be very annoying. If there's a way to help make the controllers resilient to this kind of bug (even if it's completely caused by the user), I think that would be desirable.